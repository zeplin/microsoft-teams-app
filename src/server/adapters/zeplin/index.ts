export * from "./zeplin";
export * from "./types";
