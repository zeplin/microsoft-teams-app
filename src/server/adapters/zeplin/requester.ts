import axios, { AxiosInstance, AxiosRequestConfig } from "axios";
import { ZeplinError } from "./ZeplinError";
import { INTERNAL_SERVER_ERROR } from "http-status-codes";

export class Requester {
    private instance: AxiosInstance;
    constructor(config: AxiosRequestConfig) {
        this.instance = axios.create(config);
    }

    async delete(url: string, config?: AxiosRequestConfig): Promise<void> {
        try {
            await this.instance.delete(url, config);
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async get<T = object>(url: string, config?: AxiosRequestConfig): Promise<T> {
        try {
            const { data } = await this.instance.get<T>(url, config);
            return data;
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async patch(url: string, data: object, config?: AxiosRequestConfig): Promise<void> {
        try {
            await this.instance.patch(url, data, config);
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async post(url: string, data: object, config?: AxiosRequestConfig): Promise<string> {
        try {
            const { data: { id } } = await this.instance.post<{ id: string }>(url, data, config);
            return id;
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }
}
