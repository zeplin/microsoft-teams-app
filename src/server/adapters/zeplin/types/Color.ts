export interface Color {
    id: string;
    created: number;
    name: string;
    r: number;
    g: number;
    b: number;
    a: number;
}
