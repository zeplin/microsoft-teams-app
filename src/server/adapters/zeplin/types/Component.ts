import { Thumbnails } from "./Thumbnails";

export interface Component {
    id: string;
    name: string;
    created: number;
    updated?: number;
    description?: string;
    image?: {
        width: number;
        height: number;
        original_url?: string;
        thumbnails?: Thumbnails;
    };
    section?: {
        id: string;
        group?: {
            id: string;
        };
    };
}
