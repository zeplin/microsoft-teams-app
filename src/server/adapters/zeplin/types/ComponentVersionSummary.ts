import { User } from "./User";
import { Thumbnails } from "./Thumbnails";

enum DestinationType {
    SCREEN = "screen",
    PREVIOUS = "previous",
}

export interface ComponentVersionSummary {
    id: string;
    created: number;
    creator?: User;
    commit?: {
        color: {
            r: number;
            g: number;
            b: number;
            a: number;
        };
        message?: string;
        author?: User;
    };
    image_url?: string;
    thumbnails?: Thumbnails;
    background_color?: {
        r: number;
        g: number;
        b: number;
        a: number;
    };
    source: string;
    width: number;
    height: number;
    density_scale: number;
    links: {
        rect: {
            x: number;
            y: number;
            width: number;
            height: number;
        };
        destination: {
            name?: string;
            type: DestinationType;
        };
    }[];
}
