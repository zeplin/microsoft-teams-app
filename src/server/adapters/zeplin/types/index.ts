export * from "./Organization";
export * from "./Project";
export * from "./Styleguide";
export * from "./RemPreferences";
export * from "./WebhookEvents";
export * from "./WebhookResourceType";
export * from "./ZeplinApplication";
export * from "./ProjectWebhook";
export * from "./StyleguideWebhook";
export * from "./User";
export * from "./WebhookStatus";
export * from "./WebhookUrlHealth";
