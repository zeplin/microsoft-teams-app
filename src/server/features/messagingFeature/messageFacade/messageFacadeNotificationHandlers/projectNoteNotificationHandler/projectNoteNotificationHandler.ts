import { NotificationHandler } from "../NotificationHandler";
import { AdaptiveCard, commonTeamsCard } from "../teamsCardTemplates";
import {
    EventPayload,
    EventType,
    ProjectContext,
    ScreenContext,
    WebhookEvent
} from "../../../messagingTypes";
import { ScreenNoteResource } from "../resources";
import { ZEPLIN_WEB_APP_BASE_URL, ZEPLIN_MAC_APP_URL_SCHEME } from "../../../../../config";

type ProjectNoteEventDescriptor = {
    type: EventType.PROJECT_NOTE;
    action: "created";
}

class ProjectNoteNotificationHandler extends NotificationHandler {
    // We want to send project note events immediately
    delay = 0;

    private getWebappURL(event: WebhookEvent<ProjectNoteEventPayload>): string {
        const {
            payload: {
                context: {
                    screen: {
                        id: screenId
                    },
                    project: {
                        id: projectId
                    }
                },
                resource: {
                    id: noteId
                }
            }
        } = event;
        const webappURL = new URL(ZEPLIN_WEB_APP_BASE_URL);
        webappURL.pathname = `project/${projectId}/screen/${screenId}`;
        webappURL.searchParams.set("did", noteId);
        return webappURL.toString();
    }

    // A unique grouping key so that it won't be grouped with any other events
    getGroupingKey(event: WebhookEvent): string {
        return event.deliveryId;
    }

    getTeamsMessage(events: WebhookEvent<ProjectNoteEventPayload>[]): AdaptiveCard {
        if (events.length > 1) {
            // TODO: Log this situation since we don't want it to happen
        }

        const [event] = events;
        const {
            payload: {
                context: {
                    project: {
                        id: projectId,
                        name: projectName
                    },
                    screen: {
                        id: screenId,
                        name: screenName
                    }
                },
                resource: {
                    id: noteId,
                    data: {
                        comments: [{ content: commentContent }]
                    }
                },
                actor: {
                    user: {
                        username
                    }
                }
            }
        } = event;
        const cardText = `**${username}** added a note on _${screenName}_ screen in _${projectName}_. üèÉ‚Äç‚ôÇ`;
        return commonTeamsCard({
            text: cardText,
            sectionText: commentContent,
            links: [{
                title: "Open in App",
                url: `${ZEPLIN_MAC_APP_URL_SCHEME}dot?pid=${projectId}&sid=${screenId}&did=${noteId}`
            }, {
                title: "Open in Web",
                url: this.getWebappURL(event)
            }]
        });
    }
}

export type ProjectNoteEventPayload = EventPayload<
    ProjectNoteEventDescriptor,
    ProjectContext & ScreenContext,
    ScreenNoteResource
>;
export const projectNoteNotificationHandler = new ProjectNoteNotificationHandler();