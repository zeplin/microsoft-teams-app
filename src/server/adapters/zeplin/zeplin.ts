import { ProjectWebhooks } from "./ProjectWebhooks";
import { Requester } from "./requester";
import { StyleguideWebhooks } from "./StyleguideWebhooks";

interface ZeplinOptions {
    url: string;
    webhookSecret: string;
}

class Zeplin {
    public projectWebhooks!: ProjectWebhooks;
    public styleguideWebhooks!: StyleguideWebhooks;

    init({ url, webhookSecret }: ZeplinOptions): void {
        const requester = new Requester({ baseURL: `${url}/v1/` });
        this.projectWebhooks = new ProjectWebhooks(requester, webhookSecret);
        this.styleguideWebhooks = new StyleguideWebhooks(requester, webhookSecret);
    }
}

export const zeplin = new Zeplin();
