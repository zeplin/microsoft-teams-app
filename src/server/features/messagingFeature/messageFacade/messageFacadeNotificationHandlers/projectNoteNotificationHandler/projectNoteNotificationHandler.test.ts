import { projectNoteNotificationHandler } from "./projectNoteNotificationHandler";
import { ProjectNoteEvent } from "../../../../../adapters/zeplin/types";

type GetDummyEventParams = {
    projectId?: string;
    projectName?: string;
    screenId?: string;
    screenName?: string;
    username?: string;
    noteId?: string;
    commentContent?: string;
}

function getDummyEvent({
    projectId = "projectId",
    projectName = "Project MS Teams Integration",
    screenId = "screenId",
    screenName = "Manage Zeplin Connector",
    username = "sertac",
    noteId = "noteId",
    commentContent = "Naptin nettin nettin naptin?"
}: GetDummyEventParams = {}): ProjectNoteEvent {
    return {
        payload: {
            action: "created",
            context: {
                screen: {
                    id: screenId,
                    name: screenName
                },
                project: {
                    id: projectId,
                    name: projectName
                }
            },
            actor: {
                user: {
                    username
                }
            },
            resource: {
                id: noteId,
                data: {
                    id: noteId,
                    comments: [{
                        content: commentContent
                    }]
                }
            }
        }
    } as ProjectNoteEvent;
}

describe("projectNoteNotificationHandler", () => {
    describe("getTeamsMessage", () => {
        it("should match snapshot", () => {
            expect(projectNoteNotificationHandler.getTeamsMessage([getDummyEvent()])).toMatchSnapshot();
        });
    });
});
