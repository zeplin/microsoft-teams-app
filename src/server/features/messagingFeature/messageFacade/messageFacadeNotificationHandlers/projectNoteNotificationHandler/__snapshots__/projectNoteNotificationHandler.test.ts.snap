// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectNoteNotificationHandler getTeamsMessage should match snapshot 1`] = `
{
  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "**sertac** added a note on _Manage Zeplin Connector_ screen in _Project MS Teams Integration_. üèÉ‚Äç‚ôÇ",
      "wrap": true,
      "spacing": "small"
    },
    {
      "type": "Container",
      "spacing": "extraLarge",
      "items": [
        {
          "type": "TextBlock",
          "text": "Naptin nettin nettin naptin?",
          "spacing": "none"
        }
      ]
    },
    {
      "type": "ActionSet",
      "spacing": "extraLarge",
      "actions": [
        {
          "type": "Action.OpenUrl",
          "title": "Open in App",
          "url": "zpl://dot?pid=projectId&sid=screenId&did=noteId"
        },
        {
          "type": "Action.OpenUrl",
          "title": "Open in Web",
          "url": "https://app.zeplin.io/project/projectId/screen/screenId?did=noteId"
        }
      ]
    }
  ]
}
`;
